storage: TestGossipHandlesReplacedNode failed in testrace

Found in https://teamcity.cockroachdb.com/viewLog.html?buildId=236125&buildTypeId=Cockroach_UnitTests

```
Test ended in panic. 
------- Stdout: -------
[...]
E170427 21:43:23.670624 67525 storage/queue.go:634  [replicate,n2,s2,r2/2:/System/{-tsd}] range requires a replication change, but lacks a quorum of live replicas (0/2)
W170427 21:43:24.323235 67890 storage/replica.go:4399  [n3,s3,r2/?:/System/{-tsd}] could not acquire lease for range gossip: range 2 was not found
W170427 21:43:24.323442 67890 storage/store.go:1337  [n3,s3,r2/?:/System/{-tsd}] could not gossip node liveness: range 2 was not found
E170427 21:43:24.876972 67525 storage/queue.go:634  [replicate,n2,s2,r2/2:/System/{-tsd}] range requires a replication change, but lacks a quorum of live replicas (0/2)
W170427 21:43:25.421630 67890 storage/replica.go:4399  [n3,s3,r2/?:/System/{-tsd}] could not acquire lease for range gossip: range 2 was not found
W170427 21:43:25.421805 67890 storage/store.go:1337  [n3,s3,r2/?:/System/{-tsd}] could not gossip node liveness: range 2 was not found
W170427 21:43:26.566704 67890 storage/replica.go:4399  [n3,s3,r2/?:/System/{-tsd}] could not acquire lease for range gossip: range 2 was not found
W170427 21:43:26.566852 67890 storage/store.go:1337  [n3,s3,r2/?:/System/{-tsd}] could not gossip node liveness: range 2 was not found
E170427 21:43:27.090567 67525 storage/queue.go:634  [replicate,n2,s2,r2/2:/System/{-tsd}] range requires a replication change, but lacks a quorum of live replicas (0/2)
W170427 21:43:27.568872 67890 storage/replica.go:4399  [n3,s3,r2/?:/System/{-tsd}] could not acquire lease for range gossip: range 2 was not found
W170427 21:43:27.569370 67890 storage/store.go:1337  [n3,s3,r2/?:/System/{-tsd}] could not gossip node liveness: range 2 was not found
W170427 21:43:28.484761 67890 storage/replica.go:4399  [n3,s3,r2/?:/System/{-tsd}] could not acquire lease for range gossip: range 2 was not found
W170427 21:43:28.484929 67890 storage/store.go:1337  [n3,s3,r2/?:/System/{-tsd}] could not gossip node liveness: range 2 was not found
panic: test timed out after 15m0s
```

cc @spencerkimball

--- a/pkg/storage/gossip_test.go
+++ b/pkg/storage/gossip_test.go
@@ -29,6 +29,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/internal/client"
 	"github.com/cockroachdb/cockroach/pkg/keys"
 	"github.com/cockroachdb/cockroach/pkg/roachpb"
+	"github.com/cockroachdb/cockroach/pkg/storage"
 	"github.com/cockroachdb/cockroach/pkg/testutils/testcluster"
 	"github.com/cockroachdb/cockroach/pkg/util"
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
@@ -156,10 +157,18 @@ func TestGossipHandlesReplacedNode(t *testing.T) {
 
 	tc := testcluster.StartTestCluster(t, 3,
 		base.TestClusterArgs{
-			ReplicationMode: base.ReplicationAuto,
+			// Use manual replication so that we can ensure the range is properly
+			// replicated to all three nodes before stopping one of them.
+			ReplicationMode: base.ReplicationManual,
 			ServerArgs:      serverArgs,
 		})
 	defer tc.Stopper().Stop(context.TODO())
+
+	// Ensure that the first range is fully replicated before moving on.
+	firstRangeKey := keys.MinKey
+	if _, err := tc.AddReplicas(firstRangeKey, tc.Target(1), tc.Target(2)); err != nil {
+		t.Fatal(err)
+	}
 
 	// Take down a node other than the first node and replace it with a new one.
 	// Replacing the first node would be better from an adversarial testing
@@ -170,6 +179,13 @@ func TestGossipHandlesReplacedNode(t *testing.T) {
 	newServerArgs.Addr = tc.Servers[oldNodeIdx].ServingAddr()
 	newServerArgs.PartOfCluster = true
 	newServerArgs.JoinAddr = tc.Servers[1].ServingAddr()
+	// We have to manually disable these because the testcluster only handles
+	// translating the ReplicationManual setting into store knobs for its
+	// initial nodes.
+	newServerArgs.Knobs.Store = &storage.StoreTestingKnobs{
+		DisableSplitQueue:     true,
+		DisableReplicateQueue: true,
+	}
 	tc.StopServer(oldNodeIdx)
 	tc.AddServer(t, newServerArgs)
 	tc.WaitForStores(t, tc.Server(1).Gossip())

--- a/pkg/testutils/testcluster/testcluster.go
+++ b/pkg/testutils/testcluster/testcluster.go
@@ -46,10 +46,11 @@ import (
 // TestCluster represents a set of TestServers. The hope is that it can be used
 // analoguous to TestServer, but with control over range replication.
 type TestCluster struct {
-	Servers []*server.TestServer
-	Conns   []*gosql.DB
-	stopper *stop.Stopper
-	mu      struct {
+	Servers         []*server.TestServer
+	Conns           []*gosql.DB
+	stopper         *stop.Stopper
+	replicationMode base.TestClusterReplicationMode
+	mu              struct {
 		syncutil.Mutex
 		serverStoppers []*stop.Stopper
 	}
@@ -138,21 +139,14 @@ func StartTestCluster(t testing.TB, nodes int, args base.TestClusterArgs) *TestC
 		t.Fatal("can't disable an individual server's queues when starting a cluster; " +
 			"the cluster controls replication")
 	}
-
-	switch args.ReplicationMode {
-	case base.ReplicationAuto:
-	case base.ReplicationManual:
-		if args.ServerArgs.Knobs.Store == nil {
-			args.ServerArgs.Knobs.Store = &storage.StoreTestingKnobs{}
-		}
-		storeKnobs := args.ServerArgs.Knobs.Store.(*storage.StoreTestingKnobs)
-		storeKnobs.DisableSplitQueue = true
-		storeKnobs.DisableReplicateQueue = true
-	default:
+	if args.ReplicationMode != base.ReplicationAuto && args.ReplicationMode != base.ReplicationManual {
 		t.Fatal("unexpected replication mode")
 	}
 
-	tc := &TestCluster{}
+	tc := &TestCluster{
+		stopper:         stop.NewStopper(),
+		replicationMode: args.ReplicationMode,
+	}
 	tc.stopper = stop.NewStopper()
 
 	for i := 0; i < nodes; i++ {
@@ -175,7 +169,7 @@ func StartTestCluster(t testing.TB, nodes int, args base.TestClusterArgs) *TestC
 
 	tc.WaitForStores(t, tc.Servers[0].Gossip())
 
-	if args.ReplicationMode == base.ReplicationAuto {
+	if tc.replicationMode == base.ReplicationAuto {
 		if err := tc.WaitForFullReplication(); err != nil {
 			t.Fatal(err)
 		}
@@ -187,6 +181,15 @@ func StartTestCluster(t testing.TB, nodes int, args base.TestClusterArgs) *TestC
 // the TestCluster.
 func (tc *TestCluster) AddServer(t testing.TB, serverArgs base.TestServerArgs) {
 	serverArgs.Stopper = stop.NewStopper()
+	if tc.replicationMode == base.ReplicationManual {
+		if serverArgs.Knobs.Store == nil {
+			serverArgs.Knobs.Store = &storage.StoreTestingKnobs{}
+		}
+		storeKnobs := serverArgs.Knobs.Store.(*storage.StoreTestingKnobs)
+		storeKnobs.DisableSplitQueue = true
+		storeKnobs.DisableReplicateQueue = true
+	}
+
 	s, conn, _ := serverutils.StartServer(t, serverArgs)
 	tc.Servers = append(tc.Servers, s.(*server.TestServer))
 	tc.Conns = append(tc.Conns, conn)

